### DISCUSSION TITLE
OAuth using websocket and popup

### DISCUSSION THREAD

**Discussion by Zehir** [directus/directus#20548] [Feature Requests]
### Summary

Instead of using the oauth redirect loop the idea is to use the websocket to fetch a unique link that the app could open in a pop-up and once the callback step is reached the popup is closed and the user is authenticated.

### Basic Example

```typescript
const client = createDirectus<Schema>('http://localhost:8055')
          .with(realtime({
            authMode: 'handshake',
          }))

        client.onWebSocket('open', () => {
          client.sendMessage({ type: 'oauth', provider:'github' })
        })

        client.onWebSocket('message', (message) => {
          console.log(message)
        })

        client.connect()
```
After the initial message the session is opened for x amount of time. Probably exposed with a `WEBSOCKETS_REST_OAUTH_TIMEOUT`
In response of the `{ type: 'oauth' }` message send a like like this :
```json
{
    'provider': 'github',
    'url': 'https://github.com/login?client_id=......'
}
```
The websocket session uuid is encoded somewhere in the `codeChallenge` and added in the link to be able to login the correct user.

Once the callback page is reached the user is authenticated with a access_token like the usual process of the classic oauth redirect loop.

### Motivation

The first problem I try to solve is this one :
https://discord.com/channels/725371605378924594/1145983963966492742/1178069517335793664

Long story short I was not able to use oauth when having a public API because of the cookies security params. And I don't want to expose the cookies to everybody with (`REFRESH_TOKEN_COOKIE_SAME_SITE=none`).
After the oauth redirect the cookies is stored on the directus domain and can't be acceded from my app that is on an other domain.

This also give more possibilty with the auth because in the case of single page app you won't need to reload the page after login/logout.

And you could use device auth workflow like scanning a qrcode and login from your phone like most TV apps does theses days.

### Detailed Design

See Basic Example

Also this could be abstracted in the JS sdk like this:
```typescript
        const auth = await client.oauthPopup()
        const auth = await client.oauthPopup({openPopup : true})
```
This line open a popup windows with the correct link and wait for the user to authenticate.
The promise can be rejected if the delay is expired or popup closed.

Quick recap :
- Open the websocket from client side
- Request a auth link
- Open the link in a popup or give it back to the sdk js call
- Wait for the user to auth
- On the callback page, close the popup
- Process the recieved code
- Authenticate the user

When thinking about this I looked how pockerbase do it and it's what this PR is all about : 
https://github.com/pocketbase/js-sdk/blob/98b5fa014a27b88a6f25d39d741b10f821d0a24a/src/services/RecordService.ts#L467
https://github.com/pocketbase/pocketbase/blob/985ab1e5b79a328b516decca3d1d7dc8edfcbba0/apis/record_auth.go#L637

### Must Have:
- Accept unauthenticated users for some time when `WEBSOCKETS_REST_AUTH` is `handshake` but deny any other commands.
- Be generic to be able to use with any oauth provider.

### Should Have:
- More env params to configure it like `WEBSOCKETS_REST_OAUTH_TIMEOUT`

### Drawbacks

Can't find any

### Alternatives

This can be achived with an extension but it's will need to duplicate a good amount the current code of the oauth2 driver and also you need to setup the `WEBSOCKETS_REST_AUTH` to public.

### Adoption Strategy

No breaking changes, easy to use.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
